### ISSUE TITLE
Error: couldn't evaluate grouping factor subject:area within model frame: try adding grouping factor to data frame explicitly if possible

### ISSUE THREAD

**Issue by tra6sdc** [lme4/lme4#635]
Hello.

When trying to fit a three level model (areas > subjects > time periods) I have an error with my large data set, 297k subjects and 14 time periods (with some time periods missing for some subjects). My PC has 32GB of RAM.
````
Error: couldn't evaluate grouping factor subject:area within model frame: try adding grouping factor to data frame explicitly if possible
````
This code re-creates the issue with a simulated data set:

````
library(lme4)

set.seed(123)
x <- runif(1e7)+
  0.01*rep(seq(1:10), 1e6)+
  0.00001*rep(seq(1, 1e6), each=10)+
  0.0001*rep(seq(1:1e3), each=1e4)
time <- rep(seq(1:10), 1e6)
subject <- rep(seq(1,1e6), each=10)
area <- rep(seq(1:1e3), each=1e4)

df<-data.frame(x, time, subject=factor(subject), area=factor(area))

# random missing
set.seed(123)
df <- df[-c(sample(nrow(df), 1e5)),]

m.fit <- lmer(x ~ time + (1|subject), data=df)
summary(m.fit)

mm.fit <- lmer(x ~ time + (1|area/subject), data=df)
summary(mm.fit)
````
If I scale the size of the data back by a factor of 10 (just have 1e6 observations) then the model estimates OK. I guess that the suggestion about the error message isn't relevant, and that my issue is just the size of the problem?

**Comment by bbolker** [comment]
Just checking: for which lines of code does the failure actually occur? (model 1, summary 1, model 2, summary 2?) (How long do the runs that work take on your system?) My guess is that this a memory allocation failure masquerading as something else.

**Comment by tra6sdc** [comment]
````
> system.time(mm.fit <- lmer(x ~ time + (1|area/subject), data=df))
Error: couldn't evaluate grouping factor subject:area within model frame: try adding grouping factor to data frame explicitly if possible
Timing stopped at: 115.9 113.5 285
````

**Comment by bbolker** [comment]
The first time I tried the more complex model it simply crashed the R session with "Killed" (this is the default behaviour for a Linux process that tries to grab too much memory). Trying again.

Warming up with the simpler model

```r
> system.time(p <- peakRAM(m.fit <- lmer(x ~ time + (1|subject), data=df)))
   user  system elapsed 
425.616   9.241 434.850 
> p
                            Function_Call Elapsed_Time_sec Total_RAM_Used_MiB
1 m.fit<-lmer(x~time+(1|subject),data=df)          433.679             1839.6
  Peak_RAM_Used_MiB
1            2910.9
```

Debugging through:  

- memory usage is at about 12G (out of 63 total) before starting R.
- jumps to to 21 G while running `lmer -> lFormula -> blist <- lapply(bars, mkBlist, fr, drop.unused.levels, reorder.vars = reorder.vars)`

**Comment by bbolker** [comment]
**tl;dr** you can work around this problem by constructing the interaction term yourself:

```r
df$i <- factor(paste(df$area, df$subject, sep="."))
mm.fit <- lmer(x ~ time + (1|area) + (1|i), data=df, verbose=1000)
```

This is slow (and the Hessian calculations do badly with large data sets, as we already knew) but is *not* particularly memory limited.

---

There are two issues here.

1. As you suggest, the error message you're getting is misleading. It comes from [here](https://github.com/lme4/lme4/blob/master/R/utilities.R#L80-L86), where we check whether a certain operation succeeds.  The test is there to detect the failure mode where something goes wrong with environment/variable lookup so that we can't find the variables we need, but an error *could* occur for other reasons - such as running out of memory.  It's a little bit hard to tune the error message because (1) there's not an easy way to detect which *particular* error occurred (e.g. see [this thread](https://r.789695.n4.nabble.com/suppress-specific-warnings-td4646939.html#a4647526) (and, there's not an easy way to detect all possible environment/lookup problems); (2) this is a hard problem to reproduce exactly, because you have to have little enough memory that an error is thrown about memory allocation failure, but no so little that the entire R session crashes (it's possible that using `ulimit` under Unix would make this easier, but ...)

2. More fundamentally, there is an issue with the way that R's `:` operator constructs interactions of unbalanced sets of factors; it tries to set up **all possible combinations** of levels of the two input factors. The code below tests three different ways of constructing an interaction: `f1:f2`, `interaction(f1, f2)`, and `paste(f1, f2)`. I think it would probably be too complicated to automatically "do the right thing" (we would be working around R's built-in machinery), but we could probably detect when a user was likely to run into problems because of level crossing (e.g. test whether the product of the levels of the two factors going into an interaction is much larger than the number of observations in the data set) and direct them to instructions about how to manage/work around the problem ...


```
            Elapsed_Time_sec Total_RAM_Used_MiB Peak_RAM_Used_MiB
colon                  1.036                  0              80.9
interaction            0.211                  0              39.1
paste                  0.002                  0               0.1
```

```r
library(peakRAM)

mkfac <- function(n1, n2, method = c("colon", "interaction", "paste")) {
  method <- match.arg(method)
  maxlen <- max(n1, n2)
  f1 <- factor(seq(n1))
  f2 <- factor(seq(n2))
  f1 <- rep(f1, length.out = maxlen)
  f2 <- rep(f2, length.out = maxlen)
  p <- peakRAM(f <- switch(method,
                           colon = droplevels(f1:f2),
                           interaction = interaction(f1,f2, drop=TRUE, sep=":"),
                           paste = factor(paste(f1, f2, sep=":")))
               )
  r <- p[-1]
  attr(r, "value") <- f
  return(r)
}

m1 <- mkfac(1000, 1000)
m2 <- mkfac(1000, 1000, "interaction")
m3 <- mkfac(1000, 1000, "paste")

rbind(colon= m1, interaction = m2, paste = m3)

identical(attr(m1, "value"), attr(m2, "value"))
f1 <- attr(m1, "value")
f3 <- attr(m3, "value")
setequal(levels(f3), levels(f1))
identical(as.character(f3), as.character(f1))
```

**Comment by bbolker** [comment]
I'm curious if you have tried the workaround, and if it worked?

**Comment by tra6sdc** [comment]
Hello, sorry I have not yet had an opportunity to apply the workaround. If you want, I am happy for you to close this issue.

**Comment by bbolker** [comment]
No, I'm going to leave it open. It would be good to see if the tests/warnings I suggested [above](https://github.com/lme4/lme4/issues/635#issuecomment-877903940) can be implemented straightforwardly. It would be even better to figure out if we can override the default behaviour of the `:` operator in a way that isn't too contrived ...

**Comment by bbolker** [comment]
Can you try `remotes::install_github("lme4/lme4@interaction_eval")` and see if it resolves this issue?

**Comment by bbolker** [comment]
Closing, feel free to re-open if necessary.
